src/content/reference/rules/components-and-hooks-must-be-pure.md

```diff
@@ -0,0 +1,282 @@
+---
+title: Components and Hooks must be pure
+---
+
+<Intro>
+[Purity](/learn/keeping-components-pure) makes your code easier to understand, debug, and allows React to automatically optimize your components and hooks correctly.
+</Intro>
+
+<InlineToc />
+
+<DeepDive>
+#### Why does render need to be pure? {/*why-does-render-need-to-be-pure*/}
+UI libraries like React take care of when your code runs for you so that your application has a great user experience. React is declarative: you tell React what to render in your component's logic, and React will figure out how best to display it to your user!
+
+When render is kept pure, React can understand how to prioritize which updates are most important for the user to see first. This is made possible because of render purity: since components don't have side effects in render, React can pause rendering components that aren't as important to update, and only come back to them later when it's needed.
+
+Concretely, this means that rendering logic can be run multiple times in a way that allows React to give your user a pleasant user experience. However, if your component has an untracked side effect – like modifying the value of a global variable during render – when React runs your rendering code again, your side effects will be triggered in a way that won't match what you want. This often leads to unexpected bugs that can degrade how your users experience your app.
+</DeepDive>
+
+---
+
+## Components must be idempotent {/*components-must-be-idempotent*/}
+
+React assumes that components always return the same output with respect to their inputs – props, state, and context. This is known as [idempotency](https://stackoverflow.com/questions/1077412/what-is-an-idempotent-operation).
+
+Put simply, idempotence means that you [always get the same result everytime](learn/keeping-components-pure) you run that piece of code.
+
+```js {3}
+function NewsFeed({ items }) {
+  // ✅ Array.filter doesn't mutate `items`
+  const filteredItems = items.filter(item => item.isDisplayed === true);
```

Comment:
Can you use a simplier example and make it interactive? Maybe something with numbers so its clear that if you input is num1 you'll always get f(num1) displayed if the component?

Reply:
Could you share more about what's not straightforward about this example? I'm not sure an interactive example would do anything here, it's meant to just illustrate that Array.filter is idempotent.